File: src/main/java/com/util/ActionUtils.java
package com.util;

import com.annotation.SocketAction;
import lombok.Data;
import lombok.experimental.Accessors;
import org.apache.commons.lang3.reflect.MethodUtils;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 执行事件
 */
public class ActionUtils {

    private static final Map<Integer, List<ActionMethod>> actionMap = new ConcurrentHashMap<>();


    public static void addAction(Object object) {
        Method[] methods = MethodUtils.getMethodsWithAnnotation(object.getClass(), SocketAction.class, true, true);
        for (Method method : methods) {
            SocketAction action = MethodUtils.getAnnotation(method, SocketAction.class, true, true);
            if (action != null) addAction(action, method, object);
        }
    }

    public static void addAction(SocketAction socketAction, Method method, Object object) {
        addAction(socketAction.value(), method, object);
    }

    public static void addAction(int value, Method method, Object object) {
        List<ActionMethod> methods;
        if (actionMap.containsKey(value)) {
            methods = actionMap.get(value);
        } else {
            methods = new ArrayList<>();
            actionMap.put(value, methods);
        }
        methods.add(new ActionMethod().setValue(value).setMethod(method).setObject(object));
    }

    /**
     * 执行方法
     * @param value 事件id
     * @param args 参数
     * @return 有注册方法被调用 返回true
     */
    public static boolean run(int value, Object... args) throws InvocationTargetException, IllegalAccessException {
        List<ActionMethod> methods = actionMap.get(value);
        boolean result = false;
        if (methods != null) {
            for (ActionMethod method : methods) {
                if(runMethod(method, args)) {
                    result = true;
                }
            }
        }
        return result;
    }

    private static boolean runMethod(ActionMethod method, Object... args) throws InvocationTargetException, IllegalAccessException {
        int len = args.length;
        int parameterCount = method.getMethod().getParameterCount();
        if (parameterCount == 0 || (parameterCount <= len && isVerify(method.getMethod().getParameterTypes(), args))) {
            method.run(args);
            return true;
        }
        return false;
    }

    private static boolean isVerify(Class<?>[] cls, Object... args) {
        boolean verify = false;
        for (Class<?> c : cls) {
            verify = false;
            for (Object arg : args) {
                if (Objects.equals(c, arg.getClass())) {
                    verify = true;
                    break;
                }
            }
            if (!verify) break;
        }
        return verify;
    }

    public static void removeAction(int value) {
        actionMap.remove(value);
    }

    public static void removeAction(int value, Method method) {
        List<ActionMethod> methods = actionMap.get(value);
        if (methods != null && method != null) {
            methods.remove(method);
        }
    }

}

@Accessors(chain = true)
@Data
class ActionMethod {

    int value;
    Method method;
    Object object;

    public void run(Object[] args) throws InvocationTargetException, IllegalAccessException {
        Class<?>[] types = method.getParameterTypes();
        Object[] obj = new Object[method.getParameterCount()];
        for (int i = 0; i < types.length; i++) {
            for (Object arg : args) {
                if (Objects.equals(types[i], arg.getClass())) {
                    obj[i] = arg;
                }
            }
        }
        method.setAccessible(true);
        method.invoke(object, obj);
        method.setAccessible(false);
    }

}

File: src/main/kotlin/com/util/ActionUtils.kt
package com.util

import com.annotation.SocketAction
import java.util.concurrent.ConcurrentHashMap
import kotlin.reflect.KFunction
import kotlin.reflect.KParameter
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.functions
import kotlin.reflect.full.starProjectedType
import kotlin.reflect.full.valueParameters
import kotlin.reflect.jvm.isAccessible

/**
 * 执行事件
 */
object ActionUtils {

    private val actionMap = ConcurrentHashMap<Int, MutableList<ActionMethod>>()


    @JvmStatic
    fun <T : Any> addAction(obj: T) {
        obj.javaClass.kotlin.functions.forEach { funs ->
            funs.findAnnotation<SocketAction>()?.let {
                addAction(it, funs, obj)
            }
        }
    }

    fun <T : Any> addAction(socketAction: SocketAction, method: KFunction<*>, obj: T) {
        addAction(socketAction.value, method, obj)
    }

    fun <T : Any> addAction(value: Int, method: KFunction<*>, obj: T) {
        actionMap.getOrPut(value) { mutableListOf() }.add(ActionMethod(value, method, obj))
    }

    /**
     * 执行方法
     * @param value 事件id
     * @param args 参数
     * @return 有注册方法被调用 返回true
     */
    fun <T : Any> run(value: Int, vararg args: T): Boolean {
        var result = false
        actionMap[value]?.forEach {
            if (runMethod(it, *args)) {
                result = true
            }
        }
        return result
    }

    fun <T : Any> runMethod(actionMethod: ActionMethod, vararg args: T): Boolean {
        val len = args.size
        actionMethod.method.typeParameters
        actionMethod.method.valueParameters
        val parameterCount = actionMethod.method.valueParameters.size
        if (parameterCount <= len) {
            if (parameterCount > 0) {
                val args2 = isVerify(actionMethod.method.valueParameters, *args) ?: return false
                actionMethod.run(args2.toTypedArray())
                return true
            }
            actionMethod.run()
            return true
        }
        return false
    }

    private fun isVerify(cls: List<KParameter>, vararg args: Any): List<*>? {
        val parameterTypes = cls.map { it.type }
        val argsList = args.toMutableList()
        val map = parameterTypes.map { type ->
            val value = argsList.find {
                it::class.starProjectedType.classifier == type.classifier
            }?.apply {
                argsList.remove(this)
            }
            value
        }

        return if (map.any { it == null }) {
            return null
        } else map
    }

    fun removeAction(value: Int) {
        actionMap.remove(value)
    }

    fun removeAction(value: Int, method: ActionMethod) {
        actionMap[value]?.remove(method)
    }
}


data class ActionMethod(
    var value: Int = 0,
    var method: KFunction<*>,
    var objIn: Any
) {
    fun run(args: Array<*>? = null) {
        val obj = args?.let { arrayOf(objIn, *it) } ?: arrayOf(objIn)
        method.isAccessible = true
        method.call(*obj)
        method.isAccessible = false
    }
}
